src/core/builders/create-form-validator.spec.ts

```diff
@@ -605,4 +605,124 @@ describe("createFormValidator", () => {
       expectField(Schema.arrayString)
     );
   });
+
+  it("array validation should fire validation for each field", async () => {
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
+        field: Schema.arrayObjectString,
+        rules: () => [x => (x.length < 3 ? "TOO_SHORT" : undefined)],
+      }),
+      validate.each({
+        field: Schema.arrayObjectString,
+        rules: () => [x => (x.str === "" ? "REQUIRED" : null)],
+      }),
+    ]);
+
+    const getValue = (field: FieldDescriptor<any>): any => {
+      switch (impl(field).__path) {
+        case "arrayObjectString":
+          return [{ str: "ok-string" }, { str: "" }];
+        case "arrayObjectString[0]":
+          return { str: "ok-string" };
+        case "arrayObjectString[1]":
+          return { str: "" };
+      }
+    };
+
+    const validation = await validate({
+      fields: [Schema.arrayObjectString],
+      getValue,
+    });
+
+    expect(validation).toEqual([
+      { field: Schema.arrayObjectString, error: "TOO_SHORT" },
+      { field: Schema.arrayObjectString.nth(0), error: null },
+      { field: Schema.arrayObjectString.nth(1), error: "REQUIRED" },
+    ]);
+  });
+
+  it("object validation should fire validation for each child", async () => {
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
+        field: Schema.object.num,
+        rules: () => [x => (x > 0 ? null : "REQUIRED")],
+      }),
+      validate({
+        field: Schema.object.str,
+        rules: () => [x => (x ? null : "REQUIRED")],
+      }),
+    ]);
+
+    const getValue = (field: FieldDescriptor<any>): any => {
+      switch (impl(field).__path) {
+        case "object":
+          return { str: "", num: 10 };
+        case "object.str":
+          return "";
+        case "object.num":
+          return 10;
+      }
+    };
+
+    const validation = await validate({ fields: [Schema.object], getValue });
+
+    expect(validation).toEqual([
+      { field: Schema.object.str, error: "REQUIRED" },
+      { field: Schema.object.num, error: null },
+    ]);
+  });
+
+  it("nested object validation should fire validation for each child without duplicates", async () => {
+    const { validate } = createFormValidator(Schema, validate => [
+      validate({
```

Comment:
how do we check lack of duplicates? 

Maybe the `getValue` should be `jest.fn()` so that te test can check how many times it was called fore each field?

Reply:
great point, also there're no duplicates in validation result